### ISSUE TITLE
Clearing composer cache creates a lot of work downstream

### ISSUE THREAD

**Issue by simesy** [govCMS/lagoon#78]
We [clear the composer cache](https://github.com/govCMS/govcms8lagoon/blob/master/.docker/Dockerfile.govcms8#L17) when we build the CLI container, but this is a lost opportunity. Most of the packages we use are fixed since our versions don't change often, so leaving the composer cache in place means subsequent builds downstream could speed up a lot by not needing to download all the sources - and we have a lot of sources. I propose we leave the composer cache in place here.

Something to be aware of is the main impact is on PaaS currently, but improvements in the way SaaS works may more and more require `composer install`, in order to follow mainstream best practices.

An alternative approach is to use gitlab CI cache, which might look like this (internal ref GOVCMS-3110).

```
cache:
  key: "$CI_PROJECT_PATH_SLUG"
  paths:
    - /home/.composer/cache
```

**Comment by simesy** [comment]
I checked to see appraoch this from paas scaffold might be a better option, but it's mostly the first composer install we want to save, as after that we mostly COPY /app

**Comment by simesy** [comment]
Composer cache is 81M. So adding a layer in docker with 81MB is not great, although it would only be in Dockerfile.cli.
The alternative currently is downloading 230 packages in every gitlab job. Sometimes twice.

**Comment by jackwrfuller** [comment]
@simesy is this issue still prevalent?

**Comment by jackwrfuller** [comment]
Will get this one reviewed to decide if its something we'd like to do.

**Comment by jackwrfuller** [comment]
Internal ticket: [GOVCMS-14466](https://osbfinance.atlassian.net/browse/GOVCMS-14466)

**Comment by jackwrfuller** [comment]
@simesy I got the team to look into this one, this is what they reported:

TLDR: not worth the effort.

## Option 1: Store composer cache in base image.

Composer cache adds 114MB to the base image (1.61GB). This could be viewed as negligible, especially in a compressed form hosted on Docker hub (~370MB altogether). However, SaaS sites do not run composer install and thus the only potential beneficiaries would be PaaS sites.

However, PaaS sites each have their own version of composer.json with package versions that likely to differ from the ones stored in base image cache. The benefit of the composer cache is uncertain at best.

Conclusion:  Bloating the base image with composer cache does not guarantee benefits for PaaS sites, and it can also be argued that in general it makes the matters worse.

## Option 2: Store composer cache in container registry.

Since PaaS sites do not necessarily share the same composer.json - there’s no shared cache trick that can be reliably employed. However, the pipeline could store the results of composer install to each project’s container registry (kind of like is currently being done with DB images), so then on next pipeline run if composer lock has not changed - a registry image can be pulled from registry instead of running composer install and re-downloading the packages.

However, tests with vanilla PaaS project showed that the mechanism of baking/storing/pulling the build cache actually adds more time to the pipeline. The upside is that composer is not downloading the packages from scratch, but the benefits of this as opposed to increased pipeline time + cache storage can be argued.

Conclusion:  There is no guaranteed benefit from storing build cache in container registry. Sure, some complex PaaS projects can arrive to a point where it would yield benefits, but arguably the caching solution needs to be tailored to each project and not baked into the default GovCMS CI.

As such, I think we will close this issue.
